File: azdo-pbi-analyzer/src/generators/test-generator.ts
import { PBIData, TestCase, TestCoverageSummary } from '../models/pbi-data.js';
import { SanityTestTemplate } from './templates/sanity-tests.js';
import { SmokeTestTemplate } from './templates/smoke-tests.js';
import { RegressionTestTemplate } from './templates/regression-tests.js';
import { E2ETestTemplate } from './templates/e2e-tests.js';

/**
 * Orchestrates test case generation from multiple templates
 */
export class TestGenerator {
    private sanityTemplate: SanityTestTemplate;
    private smokeTemplate: SmokeTestTemplate;
    private regressionTemplate: RegressionTestTemplate;
    private e2eTemplate: E2ETestTemplate;

    constructor() {
        this.sanityTemplate = new SanityTestTemplate();
        this.smokeTemplate = new SmokeTestTemplate();
        this.regressionTemplate = new RegressionTestTemplate();
        this.e2eTemplate = new E2ETestTemplate();
    }

    /**
     * Generate comprehensive test suite for a PBI
     */
    public generateTestSuite(
        pbi: PBIData,
        requirements: string[],
        deliverables: string[],
        isCritical: boolean
    ): { testCases: TestCase[]; coverage: TestCoverageSummary } {
        const testCases: TestCase[] = [];

        // Generate sanity tests (always)
        const sanityTests = this.sanityTemplate.generate(pbi, requirements);
        testCases.push(...sanityTests);

        // Generate smoke tests (for critical PBIs or based on priority)
        const smokeTests = this.smokeTemplate.generate(pbi, isCritical);
        testCases.push(...smokeTests);

        // Generate regression tests (always)
        const regressionTests = this.regressionTemplate.generate(pbi, requirements);
        testCases.push(...regressionTests);

        // Generate E2E tests (always)
        const e2eTests = this.e2eTemplate.generate(pbi, deliverables);
        testCases.push(...e2eTests);

        // Calculate coverage summary
        const coverage = this.calculateCoverage(testCases);

        return { testCases, coverage };
    }

    /**
     * Calculate test coverage summary
     */
    private calculateCoverage(testCases: TestCase[]): TestCoverageSummary {
        const byCategoryCount: Record<string, number> = {
            sanity: 0,
            smoke: 0,
            regression: 0,
            e2e: 0,
            integration: 0,
        };

        let automationCandidates = 0;

        testCases.forEach(test => {
            byCategoryCount[test.category] = (byCategoryCount[test.category] || 0) + 1;
            if (test.automationCandidate) {
                automationCandidates++;
            }
        });

        return {
            totalGenerated: testCases.length,
            byCategoryCount: byCategoryCount as any,
            existingTestsFound: 0, // Will be updated by analyzer
            coverageGap: 0, // Will be calculated by analyzer
            automationCandidates,
        };
    }

    /**
     * Tag tests based on priority and type
     */
    public tagTests(testCases: TestCase[], pbi: PBIData): TestCase[] {
        return testCases.map(test => {
            const tags = [...test.tags];

            // Add priority-based tags
            if (pbi.priority === 1) {
                tags.push('p1');
            }

            // Add functional area tags
            if (pbi.tags.length > 0) {
                tags.push(...pbi.tags.slice(0, 2)); // Add first 2 PBI tags
            }

            // Add area path as tag
            const areaPathParts = pbi.areaPath.split('\\');
            if (areaPathParts.length > 0) {
                const area = areaPathParts[areaPathParts.length - 1];
                tags.push(area.toLowerCase().replace(/\s+/g, '-'));
            }

            // Mark as automation candidate if criteria met
            if (test.category === 'smoke' || test.category === 'sanity') {
                if (!tags.includes('automated')) {
                    tags.push('automation-ready');
                }
            }

            return {
                ...test,
                tags: [...new Set(tags)], // Remove duplicates
            };
        });
    }
}

File: azdo-pbi-analyzer/src/AzDoPbiAnalyzer.Core/Generators/TestGenerator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using AzDoPbiAnalyzer.Core.Models;

namespace AzDoPbiAnalyzer.Core.Generators;

public class TestGenerator
{
    private readonly SanityTestTemplate _sanityTemplate = new();
    private readonly SmokeTestTemplate _smokeTemplate = new();
    private readonly RegressionTestTemplate _regressionTemplate = new();
    private readonly E2ETestTemplate _e2eTemplate = new();

    public (List<TestCase> TestCases, TestCoverageSummary Coverage) GenerateTestSuite(
        PBIData pbi, 
        List<string> requirements, 
        List<string> deliverables, 
        bool isCritical)
    {
        var testCases = new List<TestCase>();

        testCases.AddRange(_sanityTemplate.Generate(pbi, requirements));
        testCases.AddRange(_smokeTemplate.Generate(pbi, isCritical));
        testCases.AddRange(_regressionTemplate.Generate(pbi, requirements));
        testCases.AddRange(_e2eTemplate.Generate(pbi, deliverables));

        var taggedTests = TagTests(testCases, pbi);
        var coverage = CalculateCoverage(taggedTests);

        return (taggedTests, coverage);
    }

    private TestCoverageSummary CalculateCoverage(List<TestCase> testCases)
    {
        var byCategory = new Dictionary<TestCategory, int>();
        foreach (TestCategory cat in Enum.GetValues(typeof(TestCategory)))
        {
            byCategory[cat] = testCases.Count(t => t.Category == cat);
        }

        return new TestCoverageSummary(
            testCases.Count,
            byCategory,
            0,
            0,
            testCases.Count(t => t.AutomationCandidate)
        );
    }

    private List<TestCase> TagTests(List<TestCase> testCases, PBIData pbi)
    {
        var result = new List<TestCase>();
        foreach (var test in testCases)
        {
            var tags = new HashSet<string>(test.Tags);

            if (pbi.Priority == 1) tags.Add("p1");
            
            foreach (var tag in pbi.Tags.Take(2)) tags.Add(tag);

            var areaParts = pbi.AreaPath.Split('\\');
            if (areaParts.Length > 0)
            {
                tags.Add(areaParts.Last().ToLower().Replace(" ", "-"));
            }

            if ((test.Category == TestCategory.Smoke || test.Category == TestCategory.Sanity) && 
                !tags.Contains("automated"))
            {
                tags.Add("automation-ready");
            }

            result.Add(test with { Tags = tags.ToList() });
        }
        return result;
    }
}

// Templates

public class SanityTestTemplate
{
    public IEnumerable<TestCase> Generate(PBIData pbi, List<string> requirements)
    {
        var list = new List<TestCase>();
        
        list.Add(new TestCase
        {
            Id = $"SANITY-{pbi.Id}-001",
            Title = $"Verify {pbi.Title} - Feature Availability",
            Category = TestCategory.Sanity,
            Priority = TestPriority.High,
            Tags = new() { "sanity", "quick-check" },
            Steps = new() 
            {
                new(1, "Navigate to feature", "Feature accessible"),
                new(2, "Verify UI elements", "Elements visible"),
                new(3, "Check console", "No errors")
            },
            ExpectedResult = "Feature available",
            EstimatedTime = "2 minutes",
            AutomationCandidate = true
        });

        // Add more logic here mirroring the TS implementation...
        return list;
    }
}

public class SmokeTestTemplate
{
    public IEnumerable<TestCase> Generate(PBIData pbi, bool isCritical)
    {
        var list = new List<TestCase>();
        
        if (isCritical || pbi.Priority == 1)
        {
            list.Add(new TestCase
            {
                Id = $"SMOKE-{pbi.Id}-001",
                Title = $"Smoke Test: {pbi.Title} - Core Functionality",
                Category = TestCategory.Smoke,
                Priority = TestPriority.Critical,
                Tags = new() { "smoke", "critical", "automated" },
                Steps = new()
                {
                    new(1, "Execute critical flow", "Flow completes"),
                    new(2, "Verify response time", "<3s")
                },
                ExpectedResult = "Critical function works",
                EstimatedTime = "1 minute",
                AutomationCandidate = true
            });
        }
        
        return list;
    }
}

public class RegressionTestTemplate
{
    public IEnumerable<TestCase> Generate(PBIData pbi, List<string> requirements)
    {
        var list = new List<TestCase>();
        
        list.Add(new TestCase
        {
            Id = $"REGRESSION-{pbi.Id}-001",
            Title = $"Regression: {pbi.Title} - Basic Regression",
            Category = TestCategory.Regression,
            Priority = TestPriority.Medium,
            Tags = new() { "regression" },
            Steps = new() { new(1, "Verify existing functionality", "Works as before") },
            ExpectedResult = "No regressions",
            EstimatedTime = "5 minutes",
            AutomationCandidate = true
        });

        return list;
    }
}

public class E2ETestTemplate
{
    public IEnumerable<TestCase> Generate(PBIData pbi, List<string> deliverables)
    {
        var list = new List<TestCase>();
        
        list.Add(new TestCase
        {
            Id = $"E2E-{pbi.Id}-001",
            Title = $"E2E: {pbi.Title} - User Journey",
            Category = TestCategory.E2E,
            Priority = TestPriority.High,
            Tags = new() { "e2e", "journey" },
            Steps = deliverables.Select((d, i) => new TestStep(i + 1, $"Complete: {d}", "Success")).ToList(),
            ExpectedResult = "Journey complete",
            EstimatedTime = "10 minutes",
            AutomationCandidate = true
        });

        return list;
    }
}
